src/NuGet.Core/NuGet.Packaging/Rules/NoRefOrLibFolderInPackageRule.cs

```diff
@@ -0,0 +1,128 @@
+            MessageFormat = messageFormat;
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader package)
+        {
+            var files = package.GetFiles();
+            return Validate(files);
+        }
+
+        internal IEnumerable<PackagingLogMessage> Validate(IEnumerable<string> files)
+        {
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            var collection = new ContentItemCollection();
+            collection.Load(files);
+
+            var libItems = GetContentForPattern(collection, managedCodeConventions.Patterns.CompileLibAssemblies);
+            var refItems = GetContentForPattern(collection, managedCodeConventions.Patterns.CompileRefAssemblies);
+            var buildItems = GetContentForPattern(collection, managedCodeConventions.Patterns.MSBuildFiles);
+
+            var libFrameworks = GetGroupFrameworks(libItems).ToArray();
+            var refFrameworks = GetGroupFrameworks(refItems).ToArray();
+            var buildFrameworks = GetGroupFrameworks(buildItems).ToArray();
+
+            if (libFrameworks.Length == 0 && refFrameworks.Length == 0)
+            {
+                //if you can't find the ref and lib folder, then find the build folder
+                if (buildFrameworks.Length != 0)
+                {
+                    //if you can find any folders other than native or any, raise an NU5127
+                    if (buildFrameworks.Any(t => (FrameworkConstants.DotNetAll.Satisfies(t) || IsValidFramework(t))
+                        && t.GetShortFolderName() != "dotnet"
+                        && t.GetShortFolderName() != "native"))
+                    {
+                        var possibleFrameworks = buildFrameworks.
+                            Where(t => t.IsSpecificFramework && t.GetShortFolderName() != "dotnet" && t.GetShortFolderName() != "native").
+                            Select(t => t.GetShortFolderName()).ToArray();
+
+                        (var tfmNames, var suggestedDirectories) = GenerateWarningString(possibleFrameworks);
+                        
+                        var issue = new List<PackagingLogMessage>();
+                        issue.Add(PackagingLogMessage.CreateWarning(string.Format(MessageFormat, tfmNames, suggestedDirectories),
+                            NuGetLogCode.NU5127));
+                        return issue;
+                    }
+                }
+            }
+
+            return Array.Empty<PackagingLogMessage>();
+        }
+
+        private (string, string) GenerateWarningString(string[] possibleFrameworks)
+        {
+            var copy = new string[possibleFrameworks.Length - 1];
+            Array.Copy(possibleFrameworks, copy, possibleFrameworks.Length - 1);
+
+            string tfmNames = possibleFrameworks.Length > 1 ? string.Join(", ", copy)
+                + ", and " + possibleFrameworks[possibleFrameworks.Length - 1] : possibleFrameworks[0];
```

Comment:
The format I've seen most often for multi-line ternary operator is

```
condition
? true statement
: false statement
```

for example
```csharp
string tfmNames = possibleFrameworks.Length > 1
    ? string.Join(", ", copy) + ", and " + possibleFrameworks[possibleFrameworks.Length - 1] 
    : possibleFrameworks[0];
```

the line below is almost like that, but the `?` and `:` are at the end of the previous line rather than the beginning of the new line. However, I think the lines above would be made easier to read as it's not currently visually clear where the condition and statements begin and end.

Reply:
You don't need an array copy.

https://docs.microsoft.com/en-us/dotnet/api/system.string.join?view=netstandard-2.0#System_String_Join_System_String_System_String___System_Int32_System_Int32_

```cs
string.Join(", ", possibleFrameworks, 0, possibleFrameworks.Count - 1);
```